File: src/main/kotlin/org/skialin/Surface.kt
package org.skialin

import org.skialin.impl.Managed
import org.skialin.impl.NativeLoader

/**
 * [Surface] is responsible for managing the pixels that a canvas draws into.
 * 
 * The pixels can be allocated either in CPU memory (a raster surface) or on the GPU. [Surface] takes care of allocating a [Canvas] that will draw into the surface. Call [Surface.canvas] to use that canvas (but don't delete it, it is owned by the surface). [Surface] always has non-zero dimensions. If there is a request for a new surface, and either of the requested dimensions are zero, then nullptr will be returned.
 * 
 * Clients should not subclass [Surface] as there is a lot of internal machinery that is not publicly accessible.
 */
class Surface private constructor(
    ptr: Long,
) : Managed(ptr, SurfaceNative::nRelease) {
    /**
     * Returns [Canvas] that draws into [Surface].
     * 
     * Subsequent calls return the same [Canvas]. [Canvas] returned is managed and owned by [Surface], and is deleted when [Surface] is deleted.
     */
    val canvas get() = Canvas(SurfaceNative.nGetCanvas(nativePtr))

    fun makeImageSnapshot(): Image? {
        val ptr = SurfaceNative.nMakeImageSnapshot(nativePtr)
        return if (ptr == 0L) null else Image(ptr)
    }

    fun makeImageSnapshotArea(area: IRect): Image? {
        val ptr = SurfaceNative.nMakeImageSnapshotArea(nativePtr, area.left, area.top, area.right, area.bottom)
        return if (ptr == 0L) null else Image(ptr)
    }

    val width: Int get() = SurfaceNative.nWidth(nativePtr)

    val height: Int get() = SurfaceNative.nHeight(nativePtr)

    val imageInfo: ImageInfo get() = ImageInfo(SurfaceNative.nImageInfo(nativePtr))

    fun notifyContentWillChange(mode: ContentChangeMode = ContentChangeMode.RETAIN) =
        SurfaceNative.nNotifyContentWillChange(nativePtr, mode.ordinal)

    fun flush() = SurfaceNative.nFlush(nativePtr)

    fun draw(
        canvas: Canvas,
        x: Float,
        y: Float,
        paint: Paint? = null,
    ) = SurfaceNative.nDraw(nativePtr, canvas.ptr, x, y, paint?.nativePtr ?: 0L)

    enum class ContentChangeMode { DISCARD, RETAIN }

    companion object {
        fun makeRasterN32Premul(
            width: Int,
            height: Int,
        ): Surface? {
            val ptr = SurfaceNative.nMakeRasterN32Premul(width, height)
            return if (ptr == 0L) null else Surface(ptr)
        }

        fun makeRaster(info: ImageInfo): Surface? {
            val ptr = SurfaceNative.nMakeRaster(info.nativePtr)
            return if (ptr == 0L) null else Surface(ptr)
        }

        fun makeRenderTarget(
            context: DirectContext,
            budgeted: Boolean,
            info: ImageInfo,
            sampleCount: Int,
            surfaceOrigin: SurfaceOrigin,
            surfaceProps: SurfaceProps? = null,
            shouldCreateWithMips: Boolean = false,
            isProtected: Boolean = false,
        ): Surface? {
            val ptr =
                SurfaceNative.nMakeRenderTarget(
                    context.nativePtr,
                    budgeted,
                    info.nativePtr,
                    sampleCount,
                    surfaceOrigin.ordinal,
                    surfaceProps?.nativePtr ?: 0L,
                    shouldCreateWithMips,
                    isProtected,
                )
            return if (ptr == 0L) null else Surface(ptr)
        }

        fun wrapBackendTexture(
            context: DirectContext,
            backendTexture: BackendTexture,
            origin: SurfaceOrigin,
            sampleCnt: Int,
            colorType: ColorType,
            colorSpace: ColorSpace? = null,
            surfaceProps: SurfaceProps? = null,
        ): Surface? {
            val ptr =
                SurfaceNative.nWrapBackendTexture(
                    context.nativePtr,
                    backendTexture.nativePtr,
                    origin.ordinal,
                    sampleCnt,
                    colorType.ordinal,
                    colorSpace?.nativePtr ?: 0L,
                    surfaceProps?.nativePtr ?: 0L,
                )
            return if (ptr == 0L) null else Surface(ptr)
        }

        fun wrapBackendRenderTarget(
            context: DirectContext,
            backendRenderTarget: BackendRenderTarget,
            origin: SurfaceOrigin,
            colorType: ColorType,
            colorSpace: ColorSpace? = null,
            surfaceProps: SurfaceProps? = null,
        ): Surface? {
            val ptr =
                SurfaceNative.nWrapBackendRenderTarget(
                    context.nativePtr,
                    backendRenderTarget.nativePtr,
                    origin.ordinal,
                    colorType.ordinal,
                    colorSpace?.nativePtr ?: 0L,
                    surfaceProps?.nativePtr ?: 0L,
                )
            return if (ptr == 0L) null else Surface(ptr)
        }

        fun makeGraphiteRenderTarget(
            recorder: GraphiteRecorder,
            info: ImageInfo,
            mipmapped: Boolean = false,
            surfaceProps: SurfaceProps? = null,
        ): Surface? {
            val ptr = SurfaceNative.nMakeGraphiteRenderTarget(recorder.nativePtr, info.nativePtr, mipmapped, surfaceProps?.nativePtr ?: 0L)
            return if (ptr == 0L) null else Surface(ptr)
        }

        fun wrapGraphiteBackendTexture(
            recorder: GraphiteRecorder,
            backendTexture: GraphiteBackendTexture,
            colorType: ColorType,
            colorSpace: ColorSpace? = null,
            surfaceProps: SurfaceProps? = null,
        ): Surface? {
            val ptr =
                SurfaceNative.nWrapGraphiteBackendTexture(
                    recorder.nativePtr,
                    backendTexture.nativePtr,
                    colorType.ordinal,
                    colorSpace?.nativePtr ?: 0L,
                    surfaceProps?.nativePtr ?: 0L,
                )
            return if (ptr == 0L) null else Surface(ptr)
        }
    }
}

private object SurfaceNative {
    init {
        NativeLoader.ensureLoaded()
    }

    external fun nMakeRasterN32Premul(
        width: Int,
        height: Int,
    ): Long

    external fun nMakeRaster(infoPtr: Long): Long

    external fun nMakeRenderTarget(
        contextPtr: Long,
        budgeted: Boolean,
        infoPtr: Long,
        sampleCount: Int,
        surfaceOrigin: Int,
        surfacePropsPtr: Long,
        shouldCreateWithMips: Boolean,
        isProtected: Boolean,
    ): Long

    external fun nWrapBackendTexture(
        contextPtr: Long,
        backendTexturePtr: Long,
        origin: Int,
        sampleCnt: Int,
        colorType: Int,
        colorSpacePtr: Long,
        surfacePropsPtr: Long,
    ): Long

    external fun nWrapBackendRenderTarget(
        contextPtr: Long,
        backendRenderTargetPtr: Long,
        origin: Int,
        colorType: Int,
        colorSpacePtr: Long,
        surfacePropsPtr: Long,
    ): Long

    external fun nMakeGraphiteRenderTarget(
        recorderPtr: Long,
        infoPtr: Long,
        mipmapped: Boolean,
        surfacePropsPtr: Long,
    ): Long

    external fun nWrapGraphiteBackendTexture(
        recorderPtr: Long,
        backendTexturePtr: Long,
        colorType: Int,
        colorSpacePtr: Long,
        surfacePropsPtr: Long,
    ): Long

    external fun nRelease(ptr: Long)

    external fun nGetCanvas(ptr: Long): Long

    external fun nMakeImageSnapshot(ptr: Long): Long

    external fun nMakeImageSnapshotArea(
        ptr: Long,
        left: Int,
        top: Int,
        right: Int,
        bottom: Int,
    ): Long

    external fun nWidth(ptr: Long): Int

    external fun nHeight(ptr: Long): Int

    external fun nImageInfo(ptr: Long): Long

    external fun nNotifyContentWillChange(
        ptr: Long,
        mode: Int,
    )

    external fun nFlush(ptr: Long)

    external fun nDraw(
        ptr: Long,
        canvasPtr: Long,
        x: Float,
        y: Float,
        paintPtr: Long,
    )
}

File: rust/skialin-jni/src/surface.rs
use jni::sys::{jboolean, jint, jlong};
use jni::JNIEnv;

use skialin_core::{
    BackendTexture, ColorSpace, ColorType, DirectContext, GraphiteBackendTexture, GraphiteRecorder, ImageInfo, Surface, SurfaceOrigin,
    SurfaceProps,
};

use crate::util::{borrow, borrow_mut, box_ptr, drop_ptr};

#[no_mangle]
pub extern "system" fn Java_org_skialin_SurfaceNative_nMakeRasterN32Premul(
    _env: JNIEnv,
    _class: jni::objects::JClass,
    width: jint,
    height: jint,
) -> jlong {
    match Surface::new_raster_n32_premul(width, height) {
        Some(surface) => box_ptr(surface),
        None => 0,
    }
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_SurfaceNative_nMakeRaster(_env: JNIEnv, _class: jni::objects::JClass, info_ptr: jlong) -> jlong {
    let info = unsafe { borrow::<ImageInfo>(info_ptr) };
    match Surface::new_raster(info) {
        Some(surface) => box_ptr(surface),
        None => 0,
    }
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_SurfaceNative_nMakeRenderTarget(
    _env: JNIEnv,
    _class: jni::objects::JClass,
    context_ptr: jlong,
    budgeted: jboolean,
    info_ptr: jlong,
    sample_count: jint,
    surface_origin: jint,
    surface_props_ptr: jlong,
    should_create_with_mips: jboolean,
    is_protected: jboolean,
) -> jlong {
    let context = unsafe { borrow_mut::<DirectContext>(context_ptr) };
    let info = unsafe { borrow::<ImageInfo>(info_ptr) };
    let origin = if surface_origin == 0 { SurfaceOrigin::TopLeft } else { SurfaceOrigin::BottomLeft };
    let surface_props = (surface_props_ptr != 0).then(|| unsafe { borrow::<SurfaceProps>(surface_props_ptr) });
    match Surface::new_render_target(
        context,
        budgeted != 0,
        info,
        sample_count,
        origin,
        surface_props,
        should_create_with_mips != 0,
        is_protected != 0,
    ) {
        Some(surface) => box_ptr(surface),
        None => 0,
    }
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_SurfaceNative_nWrapBackendTexture(
    _env: JNIEnv,
    _class: jni::objects::JClass,
    context_ptr: jlong,
    backend_texture_ptr: jlong,
    origin: jint,
    sample_cnt: jint,
    color_type: jint,
    color_space_ptr: jlong,
    surface_props_ptr: jlong,
) -> jlong {
    let context = unsafe { borrow_mut::<DirectContext>(context_ptr) };
    let backend_texture = unsafe { borrow::<BackendTexture>(backend_texture_ptr) };
    let origin = if origin == 0 { SurfaceOrigin::TopLeft } else { SurfaceOrigin::BottomLeft };
    let color_type = crate::color_type::color_type_from_ordinal(color_type);
    let color_space = (color_space_ptr != 0).then(|| unsafe { borrow::<ColorSpace>(color_space_ptr) });
    let surface_props = (surface_props_ptr != 0).then(|| unsafe { borrow::<SurfaceProps>(surface_props_ptr) });
    match Surface::wrap_backend_texture(context, backend_texture, origin, sample_cnt, color_type, color_space, surface_props) {
        Some(surface) => box_ptr(surface),
        None => 0,
    }
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_SurfaceNative_nMakeGraphiteRenderTarget(
    _env: JNIEnv,
    _class: jni::objects::JClass,
    recorder_ptr: jlong,
    info_ptr: jlong,
    mipmapped: jboolean,
    surface_props_ptr: jlong,
) -> jlong {
    let recorder = unsafe { borrow_mut::<GraphiteRecorder>(recorder_ptr) };
    let info = unsafe { borrow::<ImageInfo>(info_ptr) };
    let surface_props = (surface_props_ptr != 0).then(|| unsafe { borrow::<SurfaceProps>(surface_props_ptr) });
    match Surface::new_graphite_render_target(recorder, info, mipmapped != 0, surface_props) {
        Some(surface) => box_ptr(surface),
        None => 0,
    }
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_SurfaceNative_nWrapGraphiteBackendTexture(
    _env: JNIEnv,
    _class: jni::objects::JClass,
    recorder_ptr: jlong,
    backend_texture_ptr: jlong,
    color_type: jint,
    color_space_ptr: jlong,
    surface_props_ptr: jlong,
) -> jlong {
    let recorder = unsafe { borrow_mut::<GraphiteRecorder>(recorder_ptr) };
    let backend_texture = unsafe { borrow::<GraphiteBackendTexture>(backend_texture_ptr) };
    let color_type = crate::color_type::color_type_from_ordinal(color_type);
    let color_space = (color_space_ptr != 0).then(|| unsafe { borrow::<ColorSpace>(color_space_ptr) });
    let surface_props = (surface_props_ptr != 0).then(|| unsafe { borrow::<SurfaceProps>(surface_props_ptr) });
    match Surface::wrap_graphite_backend_texture(recorder, backend_texture, color_type, color_space, surface_props) {
        Some(surface) => box_ptr(surface),
        None => 0,
    }
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_SurfaceNative_nRelease(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong) {
    unsafe { drop_ptr::<Surface>(ptr) };
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_SurfaceNative_nGetCanvas(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong) -> jlong {
    unsafe { borrow_mut::<Surface>(ptr) }.canvas().as_raw() as jlong
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_SurfaceNative_nMakeImageSnapshot(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong) -> jlong {
    match unsafe { borrow_mut::<Surface>(ptr) }.image_snapshot() {
        Some(image) => box_ptr(image),
        None => 0,
    }
}
